pkgs/stdenv/generic/make-derivation.nix

```diff
@@ -613,13 +613,16 @@ extendDerivation
        _derivation_original_args = derivationArg.args;
 
        builder = stdenv.shell;
-       # The bash builtin `export` dumps all current environment variables,
+       # The builtin `declare -p` dumps all bash and environment variables,
        # which is where all build input references end up (e.g. $PATH for
        # binaries). By writing this to $out, Nix can find and register
        # them as runtime dependencies (since Nix greps for store paths
        # through $out to find them)
        args = [ "-c" ''
-         export > $out
+         # this works with and without structuredAttrs
+         out="${placeholder "out"}"
+         if [ -e "$NIX_ATTRS_SH_FILE" ]; then . "$NIX_ATTRS_SH_FILE"; elif [ -f .attrs.sh ]; then . .attrs.sh; fi
```

Comment:
Probably someone depends on the format being simple environment variables on lines, which will break.
That's not necessarily a problem for undocumented behavior, but perhaps we could ease the transition by only doing `declare -p` for structured attrs, ie where it's actually needed.
Should be matter of moving this line into the conditional, and adding the original `export > $out` to a new `else` clause.